modules/management-console-agent/src/main/java/org/apache/ignite/agent/Agent.java

```diff
@@ -0,0 +1,451 @@
+        }
+        finally {
+            ctx.cache().context().database().checkpointReadUnlock();
+        }
+    }
+
+    /**
+     * Session handler for sending cluster info to backend.
+     */
+    private class AfterConnectedSessionHandler extends StompSessionHandlerAdapter {
+        /** {@inheritDoc} */
+        @Override public void afterConnected(StompSession ses, StompHeaders stompHeaders) {
+            IgniteClusterImpl cluster = ctx.cluster().get();
+
+            U.quietAndInfo(log, "");
+            U.quietAndInfo(log, "Found Management Console that can be used to monitor your cluster:: " + curSrvUri);
+
+            U.quietAndInfo(log, "");
+            U.quietAndInfo(log, "Open link in browser to monitor your cluster: " +
+                    monitoringUri(curSrvUri, cluster.id()));
+
+            U.quietAndInfo(log, "If you already using Management Console, you can add cluster manually by it's ID: " + cluster.id());
+
+            clusterSrvc.sendInitialState();
+            cacheSrvc.sendInitialState();
+
+            ses.subscribe(buildMetricsPullTopic(), new StompFrameHandler() {
+                /** {@inheritDoc} */
+                @Override public Type getPayloadType(StompHeaders headers) {
+                    return String.class;
+                }
+
+                /** {@inheritDoc} */
+                @Override public void handleFrame(StompHeaders headers, Object payload) {
+                    metricSrvc.broadcastPullMetrics();
+                }
+            });
+
+            ses.subscribe(buildActionRequestTopic(cluster.id()), new StompFrameHandler() {
+                /** {@inheritDoc} */
+                @Override public Type getPayloadType(StompHeaders headers) {
+                    return Request.class;
+                }
+
+                /** {@inheritDoc} */
+                @Override public void handleFrame(StompHeaders headers, Object payload) {
+                    actSrvc.onActionRequest((Request) payload);
+                }
+            });
+
+            cfg.setConsoleUris(singletonList(curSrvUri));
+
+            writeToMetaStorage(cfg);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void handleException(StompSession ses, StompCommand cmd, StompHeaders headers, byte[] payload, Throwable e) {
+            log.warning("Failed to process a STOMP frame", e);
+        }
+
+        /** {@inheritDoc} */
+        @Override public void handleTransportError(StompSession stompSes, Throwable e) {
+            if (e instanceof ConnectionLostException) {
+                if (disconnected.compareAndSet(false, true)) {
+                    log.error("Lost websocket connection with server: " + curSrvUri);
+
+                    reconnect();
+                }
+            }
+        }
+    }
+
+    /**
+     * Submit a reconnection task only if there no active connect in progress.
+     */
+    private void reconnect() {
+        if (connectPool.getActiveCount() == 0)
```

Comment:
Bad idea. It would be better to introduce special flag instead of useing `getActiveCount`

Reply:
@nva fixed it.